pkg/registry/pod/etcd/etcd.go

```diff
@@ -203,12 +215,31 @@ func (r *BindingREST) assignPod(ctx api.Context, podID string, machine string) e
 		return boundPodList, nil
 	})
 	if err != nil {
-		// Put the pod's host back the way it was. This is a terrible hack, but
-		// can't really be helped, since there's not really a way to do atomic
-		// multi-object changes in etcd.
-		if _, err2 := r.setPodHostTo(ctx, podID, machine, ""); err2 != nil {
+		glog.V(4).Infof("failing pod %v because constraint checks failed", podID)
+		podKey, err2 := r.store.KeyFunc(ctx, podID)
+		if err2 == nil {
+			err2 = r.store.Helper.AtomicUpdate(podKey, &api.Pod{}, false, func(obj runtime.Object) (runtime.Object, error) {
```

Comment:
I'll consolidate. I thought it would be uglier adding an extra `failed` or `rollback` parameter to `setPodHostAndAnnotations`, but it would keep the pod host & status update code all in one place -- which I think is what you're aiming for, yes?